### PR TITLE
feat: Add endpoints for managing user roles

### PR DESCRIPTION
**Pull request by dcramer**
- Add endpoints for creating, updating, deleting user roles.
- Add endpoints for assigning and unassigning roles from users.
- Add permission validation to endpoints (permissions now must be known).
- Add sudo/superuser requirements to various endpoints involving permissions.
- Add standard audit logs to all permission related endpoints.

Additionally this cleans up the various permission endpoints, improving testing, adding various security (sudo, superuser-only).

### CODE REVIEW

File: src/sentry/api/endpoints/user_role_details.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            return self.respond(status=403)
+
+        try:
+            role = UserRole.objects.get(user=user, name=role_name)
+        except UserRole.DoesNotExist:
+            return self.respond({"detail": f"'{role_name}' is not a known role."}, status=404)
+
+        with transaction.atomic():
+            with transaction.atomic():
+                deleted, _ = role.users.remove(user)
```
Comment (dcramer) [inline src/sentry/api/endpoints/user_role_details.py:55]: need to verify that m2m.remove actually gives signal here

**Comment by dcramer** [comment]
realized im missing POST/DELETE tests on users/me/roles/ - will get those added but feel free to review the actual code

**Comment by dcramer** [comment]
FYI the plan right now is to fully expose and transition to UserRole, and then remove UserPermission entirely. It will be easier to manage role-based access at the end of the day, and unless we have a viable use case to allow one-off permissions, its much simpler to couple it to roles which can be more easily managed and monitored.

File: src/sentry/api/endpoints/user_role_details.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        if not request.access.has_permission("users.admin"):
+            return self.respond(status=403)
+
+        try:
+            role = UserRole.objects.get(user=user, name=role_name)
+        except UserRole.DoesNotExist:
+            return self.respond({"detail": f"'{role_name}' is not a known role."}, status=404)
+
+        with transaction.atomic():
+            with transaction.atomic():
```
Comment (JoshFerge) [inline src/sentry/api/endpoints/user_role_details.py:65]: are the double `transaction.atomic()`'s intentional here?

Comment (dcramer) [reply src/sentry/api/endpoints/user_role_details.py:65]: nope good catch! will fix monday

File: src/sentry/api/endpoints/userroles_details.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    def delete(self, request: Request, role_name) -> Response:
+        if not request.access.has_permission("users.admin"):
+            return self.respond(status=403)
+
+        try:
+            role = UserRole.objects.get(name=role_name)
+        except UserRole.DoesNotExist:
+            return self.respond({"detail": f"'{role_name}' is not a known role."}, status=404)
+
+        with transaction.atomic():
```
Comment (JoshFerge) [inline src/sentry/api/endpoints/userroles_details.py:80]: another nested transaction.atomic ^^

**Review by JoshFerge — APPROVED** [review]
looks good overall. wondering if we could create a permission class for these endpoints as to not have the `users.admin` permission check duped in each one.

**Comment by dcramer** [comment]
@JoshFerge there is probably an opportunity to abstract permissions for sure. i didnt want to do that yet as its unclear the variability they will need (e.g. at the very least they need to be per-METHOD), but for sure something that can be done in the future.

**Pull request merged**